install/terraform/modules/gke/cluster.tf

```diff
@@ -19,37 +19,57 @@ terraform {
 
 data "google_client_config" "default" {}
 
+# A list of all parameters used in interpolation var.cluster
+# Set values to default if not key was not set in original map
+locals {
+  project           = lookup(var.cluster, "project", "agones")
+  zone              = lookup(var.cluster, "zone", "us-west1-c")
+  name              = lookup(var.cluster, "name", "test-cluster")
+  machineType       = lookup(var.cluster, "machineType", "n1-standard-4")
+  initialNodeCount  = lookup(var.cluster, "initialNodeCount", "4")
+  network           = lookup(var.cluster, "network", "default")
+  kubernetesVersion = lookup(var.cluster, "kubernetesVersion", "1.15")
+}
+
+data "google_container_engine_versions" "kubernetes" {
+  project        = local.project
+  location       = local.zone
+  version_prefix = local.kubernetesVersion
+}
+
 # echo command used for debugging purpose
 # Run `terraform taint null_resource.test-setting-variables` before second execution
 resource "null_resource" "test-setting-variables" {
   provisioner "local-exec" {
     command = <<EOT
     ${format("echo Current variables set as following - name: %s, project: %s, machineType: %s, initialNodeCount: %s, network: %s, zone: %s",
-    var.cluster["name"], var.cluster["project"],
-    var.cluster["machineType"], var.cluster["initialNodeCount"], var.cluster["network"],
-    var.cluster["zone"])}
+    local.name, local.project,
+    local.machineType, local.initialNodeCount, local.network,
+local.zone)}
     EOT
-  }
+}
 }
 
 resource "google_container_cluster" "primary" {
-  name     = var.cluster["name"]
-  location = var.cluster["zone"]
-  project  = var.cluster["project"]
-  network  = var.cluster["network"]
+  name     = local.name
+  location = local.zone
+  project  = local.project
+  network  = local.network
+
+  min_master_version = data.google_container_engine_versions.kubernetes.latest_master_version
```

Comment:
So I'm looking this over, and I'm wondering why we are doing this? Why not lean on the Google Cloud's defaulting of the minor version of the major version for Kubernetes?

It might be better to stick to what is the current minimum version default, rather than what is latest, since that will be the most stable. If people want the absolute latest, then they can opt in themselves.

@roberthbailey do you have thoughts on this?

Reply:
In this design user could use a specific version by setting `var.kubernetesVersion=1.15.2`, for instance.
There are also two other parameters in this data source which could be used to set another version:
https://www.terraform.io/docs/providers/google/d/google_container_engine_versions.html#default_cluster_version
and `valid_master_versions`. We can use smallest out of the list.

Reply:
Similar issue I found, you can see all parameters and list of versions returned:
https://github.com/hashicorp/terraform/issues/20761

Reply:
My point is, if I provide "1.15", GKE is already going to give us the default version for 1.15.x -- why are we being opinionated and forcing it to the latest 1.15.x version, when it may not be the default.

Do we have evidence that that is the best version? Most supported? most stable? Seems like the safer bet to stick with whatever GKE decides is the default, since they do the most testing.

Unless I'm missing what is happening here?

Reply:
Well the reason behind this is that `default_cluster_version` is `1.14.10-gke.27` and does not correlate to prefix which was selected:
https://gist.github.com/aLekSer/a7ac5c7728c848f3329de4caacc7c712

Reply:
For valid node versions for `us-west1-c` there exist next versions:
```
gcloud container get-server-config --zone us-west1-c
validNodeVersions:                                                                                                                                                                                         
- 1.16.8-gke.12
- 1.16.8-gke.9                                                                                                                                                                                             
- 1.15.11-gke.12
...
```
However `prefix` in `data.google_container_engine_versions.kubernetes.latest_master_version` would select only one (latest) following the prefix rule and which would be enabled for each zone in a region (if location is a region). (Currently could not see a difference in versions which are enabled on different regions). Which would be `1.15.11-gke.12`. 
I can roll back using this new `data` source here and move it into a separate PR which might be easier to review later.

Reply:
Also an elegant on how we can achieve what was before and what introduced by `data.google_container_engine_versions`:
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/d4e7dc6a585770474ea4cdee3452cf98b404c6e2/modules/beta-private-cluster/main.tf#L41
```
 master_version_regional = var.kubernetes_version != "latest" ? var.kubernetes_version : data.google_container_engine_versions.region.latest_master_version
```

Reply:
Rolled back this change.